tensorboard/defs/web.bzl

```diff
@@ -14,259 +14,295 @@
+        bind = "localhost:6006",
+        manifest = [long_path(ctx, man) for man in devserver_manifests.to_list()],
+        external_asset = [
+            struct(webpath = k, path = v)
+            for k, v in ctx.attr.external_assets.items()
+        ],
+    )
+    params_file = _new_file(ctx, "-params.pbtxt")
+    ctx.actions.write(output = params_file, content = proto.encode_text(params))
+    ctx.actions.write(
+        is_executable = True,
+        output = ctx.outputs.executable,
+        content = "#!/bin/sh\nexec %s %s" % (
+            ctx.executable._WebfilesServer.short_path,
+            long_path(ctx, params_file),
         ),
     )
-  params = struct(
-      label=str(ctx.label),
-      bind="localhost:6006",
-      manifest=[long_path(ctx, man) for man in devserver_manifests.to_list()],
-      external_asset=[struct(webpath=k, path=v)
-                      for k, v in ctx.attr.external_assets.items()])
-  params_file = _new_file(ctx, "-params.pbtxt")
-  ctx.actions.write(output=params_file, content=proto.encode_text(params))
-  ctx.actions.write(
-      is_executable=True,
-      output=ctx.outputs.executable,
-      content="#!/bin/sh\nexec %s %s" % (
-          ctx.executable._WebfilesServer.short_path,
-          long_path(ctx, params_file)))
 
-  # Export data to parent rules. This uses the legacy, string-based
-  # provider mechanism for compatibility with the base `web_library`
-  # rule from rules_closure: because `tf_web_library`s may depend on
-  # either other `tf_web_library`s or base `web_library`s, the
-  # interfaces ~must be the same.
-  #
-  # buildozer: disable=rule-impl-return
-  return struct(
-      files=depset(web_srcs + [dummy]),
-      exports=unfurl(ctx.attr.exports),
-      webfiles=struct(
-          manifest=manifest,
-          manifests=manifests,
-          webpaths=webpaths,
-          dummy=dummy,
-      ),
-      closure_js_library=collect_js(
-          unfurl(ctx.attr.deps, provider="closure_js_library"),
-          ctx.files._closure_library_base),
-      runfiles=ctx.runfiles(
-          files=(ctx.files.srcs +
-                 ctx.files.data +
-                 ctx.files._closure_library_base + [
-                     manifest,
-                     params_file,
-                     ctx.outputs.executable,
-                     dummy]),
-          transitive_files=depset(transitive=[
-              collect_runfiles([ctx.attr._WebfilesServer]),
-              collect_runfiles(deps),
-              collect_runfiles(export_deps),
-              collect_runfiles(ctx.attr.data),
-          ])))
+    # Export data to parent rules. This uses the legacy, string-based
+    # provider mechanism for compatibility with the base `web_library`
+    # rule from rules_closure: because `tf_web_library`s may depend on
+    # either other `tf_web_library`s or base `web_library`s, the
+    # interfaces ~must be the same.
+    return [
+        WebFilesInfo(
+            manifest = manifest,
```

Comment:
Just checking, are both of these expected? manifest and manifests (one plural, one singular)?

Reply:
This is what's in your code at the moment: https://github.com/tensorflow/tensorboard/blob/master/tensorboard/defs/web.bzl#L126-L127